src/abstractions/src/functions/KernelFunction.ts

```diff
@@ -1,110 +1,94 @@
-import { PromptExecutionSettings } from '../AI';
+import { defaultServiceId } from '../AI';
 import { Kernel } from '../Kernel';
-import { FromSchema, JsonSchema } from '../jsonSchema';
-import { KernelArguments } from './KernelArguments';
-
-export type Fn<Result, Args> = (args: Args, kernel?: Kernel) => Result;
-
-export type FunctionResult<
-  Schema extends JsonSchema | unknown | undefined = unknown,
-  Result = unknown,
-  Args = Schema extends JsonSchema
-    ? FromSchema<Schema>
-    : Schema extends undefined
-      ? undefined
-      : Record<string, unknown>,
-> = {
-  function?: KernelFunction<Schema, Result, Args>;
-  value?: Result;
-  renderedPrompt?: string;
-  metadata?: ReadonlyMap<string, unknown>;
-};
+import { ChatOptions } from '../chatCompletion';
+import { FromSchema } from '../jsonSchema';
+import { AIFunction } from './AIFunction';
+import { AIFunctionMetadata } from './AIFunctionMetadata';
+import { AIFunctionParameterMetadata } from './AIFunctionParameterMetadata';
+import { FunctionName } from './FunctionName';
+
+
+// export type Fn<Result, Args> = (args: Args, kernel?: Kernel) => Result;
+// 
+// export type FunctionResult<
+//   Schema extends JsonSchema | unknown | undefined = unknown,
+//   Result = unknown,
+//   Args = Schema extends JsonSchema
+//     ? FromSchema<Schema>
+//     : Schema extends undefined
+//       ? undefined
+//       : Record<string, unknown>,
+// > = {
+//   function?: KernelFunction<Schema, Result, Args>;
+//   value?: Result;
+//   renderedPrompt?: string;
+//   metadata?: ReadonlyMap<string, unknown>;
+// };
 
 export type KernelFunctionProps<Props> = Props;
 
-export type KernelFunctionMetadata<Schema extends JsonSchema | unknown> = {
-  name: string;
-  description?: string;
+export class KernelFunctionMetadata<PARAMETERS = AIFunctionParameterMetadata> extends AIFunctionMetadata<PARAMETERS> {
```

Comment:
nit: why is the PARAMETERS in caps, usually reserved for constants

Reply:
I believe all caps is what people usually use for generic types, but yeah I agree it looks a bit weird. I will rename